### PR TITLE
Fix grpc map field serialization issue and optional field deserialization issue(master)

### PR DESCRIPTION
**Pull request by daneshk**
## Purpose
Fix grpc map field serialization issue and an optional field deserialization issue

Resolves: #20888
Resolves: #20987

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-integration-test/src/test/resources/grpc/src/clients/19_grpc_map_field_client.bal [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    io:println(testMapFields());
+    io:println(testOptionalFields());
+}
+
+public function testMapFields() returns string {
+    MetricsPublishRequest request = {id: "xxxxx", metrics: [{timestamp:1580966325916,
+    name:"ballerina/http/Caller_3XX_requests_total_count",value:0.0,tags:[{key:"action", value:"respond"}]}]};
+    grpc:Headers|error publishMetrics = blockingEp->publishMetrics(request);
+    if (publishMetrics is error) {
+        return io:sprintf("Metrics publish failed: %s - %s", publishMetrics.reason(), <string> publishMetrics.detail()["message"]);
```
Comment (anupama-pathirage) [inline tests/jballerina-integration-test/src/test/resources/grpc/src/clients/19_grpc_map_field_client.bal]: Shall we stick to the 120 max line width? Check all the places.

File: tests/jballerina-integration-test/src/test/resources/grpc/src/grpcservices/19_grpc_map_service.bal [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+public type MetricsPublishRequest record {|
+    string id = "";
+    Metric[] metrics = [];
+|};
+
+public type Metric record {|
+    int timestamp = 0;
+    string name = "";
+    float value = 0.0;
+    record {| string key; string value; |}[] tags = [];
```
Comment (anupama-pathirage) [inline tests/jballerina-integration-test/src/test/resources/grpc/src/grpcservices/19_grpc_map_service.bal]:
```suggestion
    record {|string key; string value;|}[] tags = [];
```

Comment (anupama-pathirage) [reply tests/jballerina-integration-test/src/test/resources/grpc/src/grpcservices/19_grpc_map_service.bal]: Shall we reformat the .bal files?

File: stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/nativeimpl/client/FunctionUtils.java [hunk 3]
```diff
@@ -248,18 +248,22 @@ public static Object externBlockingExecute(ObjectValue clientEndpoint, String me
...
+                    dataContext = new DataContext(Scheduler.getStrand(),
                             new NonBlockingCallback(Scheduler.getStrand()));
                     blockingStub.executeUnary(requestMsg, methodDescriptors.get(methodName), dataContext);
                 } else {
                     return notifyErrorReply(INTERNAL, "Error while executing the client call. Method type " +
                             methodType.name() + " not supported");
                 }
             } catch (Exception e) {
+                if (dataContext != null) {
+                    dataContext.getCallback().notifyFailure(MessageUtils.getConnectorError(e));
```
Comment (chamil321) [inline stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/nativeimpl/client/FunctionUtils.java]: What is the special reason to notify the error to callback and return error subsequently?
If the exception has occurred before the non-blocking call, then call back should not be called ideally. We can simply unblock strand[1] and return error. Problem is if we notify the callback when the non-blocking task is not processed, that can cause strand state issues.

[1] https://github.com/ballerina-platform/ballerina-lang/blob/master/stdlib/http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/Respond.java#L122

Comment (daneshk) [reply stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/nativeimpl/client/FunctionUtils.java]: We need to release the strand when there is an error before doing the network call. Thanks for the suggestion

Comment (daneshk) [reply stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/nativeimpl/client/FunctionUtils.java]: Fixed in 25dc241

**Pull request merged**